File: backend/engine/pda_engine.py
from typing import List, Optional, Set
try:
    from models.schemas import PDAMachine, SimulationResult, SimulationStepTrace
except ImportError:
    from ..models.schemas import PDAMachine, SimulationResult, SimulationStepTrace

def is_epsilon(sym: Optional[str]) -> bool:
    return not sym or sym in ('ε', 'eps', '', 'lambda')

def tokenize_stack_symbols(s: str, alphabet: List[str]) -> List[str]:
    if not s or is_epsilon(s):
        return []
    sorted_alpha = sorted([a for a in alphabet if a and not is_epsilon(a)], key=lambda x: len(x), reverse=True)
    tokens = []
    remaining = s
    while remaining:
        match = next((a for a in sorted_alpha if remaining.startswith(a)), None)
        if match:
            tokens.append(match)
            remaining = remaining[len(match):]
        else:
            tokens.append(remaining[0])
            remaining = remaining[1:]
    return tokens

class PDAConfig:
    def __init__(self, state: str, input_index: int, stack: List[str], trace: List[SimulationStepTrace]):
        self.state = state
        self.input_index = input_index
        self.stack = stack
        self.trace = trace

def simulate_pda(machine: PDAMachine, input_str: str, max_steps: int = 500) -> SimulationResult:
    init_stack = [machine.initialStackSymbol] if machine.initialStackSymbol else ['Z0']
    start_state = machine.startState
    stack_alphabet = machine.stackAlphabet or ['Z0', 'a', 'b', 'c', '0', '1', '(', ')']

    initial_trace = SimulationStepTrace(
        stepIndex=0,
        activeStates=[start_state],
        remainingInput=input_str,
        readSymbol=input_str[0] if len(input_str) > 0 else None,
        activeTransitionId=None,
        stack=list(init_stack),
        actionSummary=f"Initial state: {start_state}, Stack: [{', '.join(init_stack)}]",
        status="RUNNING",
    )

    def is_accepted_config(cfg: PDAConfig) -> bool:
        input_consumed = cfg.input_index >= len(input_str)
        if not input_consumed:
            return False
        if machine.acceptanceMode == 'empty_stack':
            return len(cfg.stack) == 0
        else:
            return cfg.state in machine.acceptStates

    queue: List[PDAConfig] = [
        PDAConfig(
            state=start_state,
            input_index=0,
            stack=list(init_stack),
            trace=[initial_trace],
        )
    ]

    best_reject_config: Optional[PDAConfig] = None
    visited: Set[str] = set()
    explored_count = 0

    while queue and explored_count < max_steps:
        explored_count += 1
        current = queue.pop(0)

        # Check acceptance
        if is_accepted_config(current):
            final_traces = [t.model_copy() for t in current.trace]
            final_traces[-1].status = "ACCEPTED"
            final_traces[-1].actionSummary += f" | Accepted via {'empty stack' if machine.acceptanceMode == 'empty_stack' else f'final state {current.state}'}"
            return SimulationResult(
                accepted=True,
                finalStatus="ACCEPTED",
                totalSteps=len(final_traces),
                traces=final_traces,
                message=f'Input string "{input_str}" ACCEPTED by PDA.',
            )

        if not best_reject_config or current.input_index > best_reject_config.input_index or len(current.trace) > len(best_reject_config.trace):
            best_reject_config = current

        # Cycle detection
        sig = f"{current.state}:{current.input_index}:{','.join(current.stack)}"
        if sig in visited:
            continue
        visited.add(sig)

        current_input_sym = input_str[current.input_index] if current.input_index < len(input_str) else ""
        top_of_stack = current.stack[-1] if len(current.stack) > 0 else ""

        for t in machine.transitions:
            if t.from_state != current.state:
                continue

            input_matches = is_epsilon(t.inputSymbol) or t.inputSymbol == current_input_sym
            stack_matches = is_epsilon(t.popSymbol) or t.popSymbol == top_of_stack

            if input_matches and stack_matches:
                reads_input = not is_epsilon(t.inputSymbol)
                next_input_index = current.input_index + 1 if reads_input else current.input_index

                next_stack = list(current.stack)
                if not is_epsilon(t.popSymbol):
                    if len(next_stack) == 0 or next_stack[-1] != t.popSymbol:
                        continue
                    next_stack.pop()

                if not is_epsilon(t.pushSymbols):
                    push_tokens = tokenize_stack_symbols(t.pushSymbols, stack_alphabet)
                    for sym in reversed(push_tokens):
                        next_stack.append(sym)

                next_remaining = input_str[next_input_index:]
                read_sym = t.inputSymbol if reads_input else 'ε'
                action_desc = f"Read '{read_sym}', Pop '{t.popSymbol or 'ε'}', Push '{t.pushSymbols or 'ε'}' -> State {t.to_state}"

                new_trace_step = SimulationStepTrace(
                    stepIndex=len(current.trace),
                    activeStates=[t.to_state],
                    remainingInput=next_remaining,
                    readSymbol=read_sym,
                    activeTransitionId=t.id,
                    stack=list(next_stack),
                    actionSummary=action_desc,
                    status="RUNNING",
                )

                queue.append(
                    PDAConfig(
                        state=t.to_state,
                        input_index=next_input_index,
                        stack=next_stack,
                        trace=current.trace + [new_trace_step],
                    )
                )

    reject_traces = [t.model_copy() for t in best_reject_config.trace] if best_reject_config else [initial_trace]
    reject_traces[-1].status = "REJECTED"
    reject_traces[-1].actionSummary += " | No valid accepting path found (Rejected)"

    return SimulationResult(
        accepted=False,
        finalStatus="REJECTED",
        totalSteps=len(reject_traces),
        traces=reject_traces,
        message=f'Input string "{input_str}" REJECTED by PDA (explored {explored_count} configurations).',
    )

File: frontend/src/engine/pdaEngine.ts
import { PDAMachine, PDATransition, SimulationResult, SimulationStepTrace } from '../types/automata';
import { isEpsilon } from './nfaEngine';

export function tokenizeStackSymbols(str: string, alphabet: string[] = []): string[] {
  if (!str || isEpsilon(str)) return [];
  const sortedAlpha = [...alphabet.filter((a) => a && !isEpsilon(a))].sort((a, b) => b.length - a.length);
  const tokens: string[] = [];
  let remaining = str;
  while (remaining.length > 0) {
    const match = sortedAlpha.find((a) => remaining.startsWith(a));
    if (match) {
      tokens.push(match);
      remaining = remaining.slice(match.length);
    } else {
      tokens.push(remaining[0]);
      remaining = remaining.slice(1);
    }
  }
  return tokens;
}

interface PDAConfig {
  state: string;
  inputIndex: number;
  stack: string[]; // stack array, last element is top of stack
  trace: SimulationStepTrace[];
}

export function simulatePDA(machine: PDAMachine, inputString: string, maxSteps = 500): SimulationResult {
  const initStack = machine.initialStackSymbol ? [machine.initialStackSymbol] : ['Z0'];
  const startState = machine.startState;

  if (!startState) {
    return {
      accepted: false,
      finalStatus: 'ERROR',
      traces: [{
        stepIndex: 0,
        activeStates: [],
        remainingInput: inputString,
        inputIndex: 0,
        stack: initStack,
        actionSummary: 'No start state defined for PDA',
        status: 'ERROR',
      }],
      totalSteps: 1,
      message: 'No start state defined for PDA.',
    };
  }

  const initialTrace: SimulationStepTrace = {
    stepIndex: 0,
    activeStates: [startState],
    currentSymbol: inputString.length > 0 ? inputString[0] : undefined,
    inputIndex: 0,
    remainingInput: inputString,
    stack: [...initStack],
    topOfStack: initStack[initStack.length - 1],
    actionSummary: `Initial state: ${startState}, Stack: [${initStack.join(', ')}]`,
    status: 'RUNNING',
  };

  const isAcceptedConfig = (cfg: PDAConfig): boolean => {
    const inputConsumed = cfg.inputIndex >= inputString.length;
    if (!inputConsumed) return false;

    if (machine.acceptanceMode === 'empty_stack') {
      return cfg.stack.length === 0;
    } else {
      // By final state
      return machine.acceptStates.includes(cfg.state);
    }
  };

  // BFS / Priority search to find accepting path or longest valid path
  const queue: PDAConfig[] = [
    {
      state: startState,
      inputIndex: 0,
      stack: [...initStack],
      trace: [initialTrace],
    },
  ];

  let bestRejectConfig: PDAConfig | null = null;
  const visited = new Set<string>();
  let exploredCount = 0;

  while (queue.length > 0 && exploredCount < maxSteps) {
    exploredCount++;
    const current = queue.shift()!;

    // Check acceptance
    if (isAcceptedConfig(current)) {
      const finalTraces = [...current.trace];
      const lastTrace = finalTraces[finalTraces.length - 1];
      lastTrace.status = 'ACCEPTED';
      lastTrace.actionSummary += ` | Accepted via ${machine.acceptanceMode === 'empty_stack' ? 'empty stack' : `final state ${current.state}`}`;

      return {
        accepted: true,
        finalStatus: 'ACCEPTED',
        traces: finalTraces,
        totalSteps: finalTraces.length,
        message: `Input string "${inputString}" ACCEPTED by PDA.`,
      };
    }

    if (!bestRejectConfig || current.inputIndex > bestRejectConfig.inputIndex || current.trace.length > bestRejectConfig.trace.length) {
      bestRejectConfig = current;
    }

    // Config signature for cycle detection
    const sig = `${current.state}:${current.inputIndex}:${current.stack.join('|')}`;
    if (visited.has(sig)) continue;
    visited.add(sig);

    // Explore possible transitions from this configuration
    const currentInputSym = current.inputIndex < inputString.length ? inputString[current.inputIndex] : '';
    const topOfStack = current.stack.length > 0 ? current.stack[current.stack.length - 1] : '';

    // Find transitions matching:
    // 1) matching inputSymbol or ε
    // 2) matching popSymbol or ε
    const possibleTransitions = machine.transitions.filter(t => {
      if (t.from !== current.state) return false;

      const inputMatches = isEpsilon(t.inputSymbol) || t.inputSymbol === currentInputSym;
      const stackMatches = isEpsilon(t.popSymbol) || t.popSymbol === topOfStack;

      return inputMatches && stackMatches;
    });

    for (const t of possibleTransitions) {
      const readsInput = !isEpsilon(t.inputSymbol);
      const nextInputIndex = readsInput ? current.inputIndex + 1 : current.inputIndex;

      // Stack manipulation
      const nextStack = [...current.stack];
      if (!isEpsilon(t.popSymbol)) {
        if (nextStack.length === 0 || nextStack[nextStack.length - 1] !== t.popSymbol) {
          continue; // Cannot pop
        }
        nextStack.pop();
      }

      // Push symbols
      if (!isEpsilon(t.pushSymbols)) {
        // Tokenize using stackAlphabet and push in reverse order so first token ends up on top
        const tokens = tokenizeStackSymbols(t.pushSymbols, machine.stackAlphabet || ['Z0', 'a', 'b', 'c', '0', '1', '(', ')']);
        for (let i = tokens.length - 1; i >= 0; i--) {
          nextStack.push(tokens[i]);
        }
      }

      const nextRemaining = inputString.slice(nextInputIndex);
      const nextSymbol = nextInputIndex < inputString.length ? inputString[nextInputIndex] : undefined;

      const actionDesc = `Read '${readsInput ? t.inputSymbol : 'ε'}', Pop '${t.popSymbol || 'ε'}', Push '${t.pushSymbols || 'ε'}' → State ${t.to}`;

      const newTraceStep: SimulationStepTrace = {
        stepIndex: current.trace.length,
        activeStates: [t.to],
        currentSymbol: nextSymbol,
        inputIndex: nextInputIndex,
        remainingInput: nextRemaining,
        stack: [...nextStack],
        topOfStack: nextStack.length > 0 ? nextStack[nextStack.length - 1] : undefined,
        activeTransitionId: t.id,
        actionSummary: actionDesc,
        status: 'RUNNING',
      };

      queue.push({
        state: t.to,
        inputIndex: nextInputIndex,
        stack: nextStack,
        trace: [...current.trace, newTraceStep],
      });
    }
  }

  // If no accepting path was found
  const rejectTraces = bestRejectConfig ? [...bestRejectConfig.trace] : [initialTrace];
  const lastRejectTrace = rejectTraces[rejectTraces.length - 1];
  lastRejectTrace.status = 'REJECTED';
  lastRejectTrace.actionSummary += ' | No valid accepting path found (Rejected)';

  return {
    accepted: false,
    finalStatus: 'REJECTED',
    traces: rejectTraces,
    totalSteps: rejectTraces.length,
    message: `Input string "${inputString}" REJECTED by PDA (explored ${exploredCount} configurations).`,
  };
}
